PR: https://github.com/odoo/odoo/pull/225953

From: a3b1bce1b3af6f521439196f9433f2a4fb6819ed
From: damrOdoo
Date: 2025-09-09 06:00:02

Breaking data model changes scores: del:4 + add:0, change matches:
-    _inherit = "project.task.template"
-    allow_timesheets = fields.Boolean(
-    encode_uom_in_days = fields.Boolean(compute='_compute_encode_uom_in_days', default=lambda self: self._uom_in_days(), export_string_translation=False)
-    display_name = fields.Char(help="""Use these keywords in the title to set new tasks:\n

Total Changes: 74

[REV] project, project_* : revert the refactoring of task.template

This commit's purpose is to revert the refactor of the task template
done in task-4919945. There should be no functional changes due to
this commit. The revert is done because the refactor adds too many
code lines and is too complex.

Task-5072878

Part-of: odoo/odoo#225953
Related: odoo/enterprise#94121
Related: odoo/upgrade#8407
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr_timesheet/models/__init__.py
+++ b/addons/hr_timesheet/models/__init__.py
@@ -10,7 +10,6 @@ from . import ir_ui_menu
 from . import res_company
 from . import res_config_settings
 from . import project_project
-from . import project_task_template
 from . import project_task
 from . import project_update
 from . import project_collaborator

--- a/addons/hr_timesheet/models/hr_timesheet.py
+++ b/addons/hr_timesheet/models/hr_timesheet.py
@@ -62,7 +62,7 @@ class AccountAnalyticLine(models.Model):
     task_id = fields.Many2one(
         'project.task', 'Task', index='btree_not_null',
         compute='_compute_task_id', store=True, readonly=False,
-        domain="[('allow_timesheets', '=', True), ('project_id', '=?', project_id)]")
+        domain="[('allow_timesheets', '=', True), ('project_id', '=?', project_id), ('has_template_ancestor', '=', False)]")
     parent_task_id = fields.Many2one('project.task', related='task_id.parent_id', store=True, index='btree_not_null')
     project_id = fields.Many2one(
         'project.project', 'Project', domain=_domain_project_id, index=True,

--- a/addons/hr_timesheet/models/project_task.py
+++ b/addons/hr_timesheet/models/project_task.py
@@ -29,8 +29,12 @@ PROJECT_TASK_READABLE_FIELDS = {
 class ProjectTask(models.Model):
     _inherit = "project.task"
 
-    project_id = fields.Many2one(domain="['|', ('company_id', '=', False), ('company_id', '=?',  company_id), ('is_internal_project', '=', False)]")
+    project_id = fields.Many2one(domain="['|', ('company_id', '=', False), ('company_id', '=?',  company_id), ('is_internal_project', '=', False), ('is_template', 'in', [is_template, False])]")
     analytic_account_active = fields.Boolean("Active Analytic Account", related='project_id.analytic_account_active', export_string_translation=False)
+    allow_timesheets = fields.Boolean(
+        "Allow timesheets",
+        compute='_compute_allow_timesheets', search='_search_allow_timesheets',
+        compute_sudo=True, readonly=True, export_string_translation=False)
     remaining_hours = fields.Float("Time Remaining", compute='_compute_remaining_hours', store=True, readonly=True, help="Number of allocated hours minus the number of hours spent.")
     remaining_hours_percentage = fields.Float(compute='_compute_remaining_hours_percentage', search='_search_remaining_hours_percentage', export_string_translation=False)
     effective_hours = fields.Float("Time Spent", compute='_compute_effective_hours', compute_sudo=True, store=True)
@@ -39,6 +43,16 @@ class ProjectTask(models.Model):
     overtime = fields.Float(compute='_compute_progress_hours', store=True)
     subtask_effective_hours = fields.Float("Time Spent on Sub-tasks", compute='_compute_subtask_effective_hours', recursive=True, store=True, help="Time spent on the sub-tasks (and their own sub-tasks) of this task.")
     timesheet_ids = fields.One2many('account.analytic.line', 'task_id', 'Timesheets', export_string_translation=False)
+    encode_uom_in_days = fields.Boolean(compute='_compute_encode_uom_in_days', default=lambda self: self._uom_in_days(), export_string_translation=False)
+    display_name = fields.Char(help="""Use these keywords in the title to set new tasks:\n
+        30h Allocate 30 hours to the task
+        #tags Set tags on the task
+        @user Assign the task to a user
+        ! Set the task a medium priority
+        !! Set the task a high priority
+        !!! Set the task a urgent priority\n
+        Make sure to use the right format and order e.g. Improve the configuration screen 5h #feature #v16 @Mitchell !""",
+    )
     @property
     def TASK_PORTAL_READABLE_FIELDS(self):
         return super().TASK_PORTAL_READABLE_FIELDS | PROJECT_TASK_READABLE_FIELDS
@@ -49,6 +63,23 @@ class ProjectTask(models.Model):
         if private_tasks and self.env['account.analytic.line'].sudo().search_count([('task_id', 'in', private_tasks.ids)], limit=1):
             raise UserError(_("This task cannot be private because there are some timesheets linked to it."))
 
+    def _uom_in_days(self):
+        return self.env.company.timesheet_encode_uom_id == self.env.ref('uom.product_uom_day')
+
+    def _compute_encode_uom_in_days(self):
+        self.encode_uom_in_days = self._uom_in_days()
+
+    @api.depends('project_id.allow_timesheets')
+    def _compute_allow_timesheets(self):
+        for task in self:
+            task.allow_timesheets = task.project_id.allow_timesheets
+
+    def _search_allow_timesheets(self, operator, value):
+        query = self.env['project.project'].sudo()._search([
+            ('allow_timesheets', operator, value),
+        ])
+        return [('project_id', 'in', query)]
+
     @api.depends('timesheet_ids.unit_amount')
     def _compute_effective_hours(self):
         if not any(self._ids):

--- a/addons/hr_timesheet/models/project_task_template.py
+++ b/None
@@ -1,38 +0,0 @@
-from odoo import models, fields, api
-
-
-class ProjectTaskTemplate(models.Model):
-    _inherit = "project.task.template"
-
-    allow_timesheets = fields.Boolean(
-        "Allow timesheets",
-        compute='_compute_allow_timesheets', search='_search_allow_timesheets',
-        compute_sudo=True, readonly=True, export_string_translation=False)
-
-    encode_uom_in_days = fields.Boolean(compute='_compute_encode_uom_in_days', default=lambda self: self._uom_in_days(), export_string_translation=False)
-    display_name = fields.Char(help="""Use these keywords in the title to set new tasks:\n
-        30h Allocate 30 hours to the task
-        #tags Set tags on the task
-        @user Assign the task to a user
-        ! Set the task a medium priority
-        !! Set the task a high priority
-        !!! Set the task a urgent priority\n
-        Make sure to use the right format and order e.g. Improve the configuration screen 5h #feature #v16 @Mitchell !""",
-    )
-
-    def _uom_in_days(self):
-        return self.env.company.timesheet_encode_uom_id == self.env.ref('uom.product_uom_day')
-
-    def _compute_encode_uom_in_days(self):
-        self.encode_uom_in_days = self._uom_in_days()
-
-    @api.depends('project_id.allow_timesheets')
-    def _compute_allow_timesheets(self):
-        for task in self:
-            task.allow_timesheets = task.project_id.allow_timesheets
-
-    def _search_allow_timesheets(self, operator, value):
-        query = self.env['project.project'].sudo()._search([
-            ('allow_timesheets', operator, value),
-        ])
-        return [('project_id', 'in', query)]
